# 语言翻译

在此项目中，你将了解神经网络机器翻译这一领域。你将用由英语和法语语句组成的数据集，训练一个序列到序列模型（sequence to sequence model），该模型能够将新的英语句子翻译成法语。

## 获取数据

因为将整个英语语言内容翻译成法语需要大量训练时间，所以我们提供了一小部分的英语语料库。


In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## 探索数据

研究 view_sentence_range，查看并熟悉该数据的不同部分。


In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## 实现预处理函数

### 文本到单词 id

和之前的 RNN 一样，你必须首先将文本转换为数字，这样计算机才能读懂。在函数 `text_to_ids()` 中，你需要将单词中的 `source_text` 和 `target_text` 转为 id。但是，你需要在 `target_text` 中每个句子的末尾，添加 `<EOS>` 单词 id。这样可以帮助神经网络预测句子应该在什么地方结束。


你可以通过以下代码获取  `<EOS> ` 单词ID：

```python
target_vocab_to_int['<EOS>']
```

你可以使用 `source_vocab_to_int` 和 `target_vocab_to_int` 获得其他单词 id。


In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    source_letter_ids = [[source_vocab_to_int.get(letter, source_vocab_to_int['<UNK>']) for letter in line.split()] for line in source_text.split('\n')]
    target_letter_ids = [[target_vocab_to_int.get(letter, target_vocab_to_int['<UNK>']) for letter in line.split()] + [target_vocab_to_int['<EOS>']] for line in target_text.split('\n')] 
    
    return source_letter_ids, target_letter_ids

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### 预处理所有数据并保存

运行以下代码单元，预处理所有数据，并保存到文件中。


In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# 检查点

这是你的第一个检查点。如果你什么时候决定再回到该记事本，或需要重新启动该记事本，可以从这里继续。预处理的数据已保存到磁盘上。

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### 检查 TensorFlow 版本，确认可访问 GPU

这一检查步骤，可以确保你使用的是正确版本的 TensorFlow，并且能够访问 GPU。


In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf

# Check TensorFlow Version
assert LooseVersion(tf.__version__) in [LooseVersion('1.0.0'), LooseVersion('1.0.1')], 'This project requires TensorFlow version 1.0  You are using {}'.format(tf.__version__)
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.0.1
Default GPU Device: /gpu:0


## 构建神经网络

你将通过实现以下函数，构建出要构建一个序列到序列模型所需的组件：

- `model_inputs`
- `process_decoding_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### 输入

实现 `model_inputs()` 函数，为神经网络创建 TF 占位符。该函数应该创建以下占位符：

- 名为 “input” 的输入文本占位符，并使用 TF Placeholder 名称参数（等级（Rank）为 2）。
- 目标占位符（等级为 2）。
- 学习速率占位符（等级为 0）。
- 名为 “keep_prob” 的保留率占位符，并使用 TF Placeholder 名称参数（等级为 0）。

在以下元祖（tuple）中返回占位符：（输入、目标、学习速率、保留率）


In [7]:
import tensorflow as tf
def model_inputs():
    """
    Create TF Placeholders for input, targets, and learning rate.
    :return: Tuple (input, targets, learning rate, keep probability)
    """
    # TODO: Implement Function
    inputs = tf.placeholder(tf.int32, [None, None], name='input')
    targets = tf.placeholder(tf.int32, [None, None], name='targets')
    learning_rate = tf.placeholder(tf.float32, name='learning_rate')
    keep_prob = tf.placeholder(tf.float32, name='keep_prob')
    
    return inputs, targets, learning_rate, keep_prob

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### 处理解码输入

使用 TensorFlow 实现 `process_decoding_input`，以便删掉 `target_data` 中每个批次的最后一个单词 ID，并将 GO ID 放到每个批次的开头。

In [8]:
def process_decoding_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for dencoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    dec_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    
    return dec_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_decoding_input(process_decoding_input)

Tests Passed


### 编码

实现 `encoding_layer()`，以使用 [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn) 创建编码器 RNN 层级。

In [9]:
def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :return: RNN state
    """
    # TODO: Implement Function
    lstm = tf.contrib.rnn.BasicLSTMCell(rnn_size)
    drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
    enc_cell = tf.contrib.rnn.MultiRNNCell([drop] * num_layers)
    _, enc_state = tf.nn.dynamic_rnn(enc_cell, rnn_inputs, dtype=tf.float32)
    
    return enc_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### 解码 - 训练

使用 [`tf.contrib.seq2seq.simple_decoder_fn_train()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/simple_decoder_fn_train) 和 [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder) 创建训练分对数（training logits）。将 `output_fn` 应用到 [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder) 输出上。

In [10]:
def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, sequence_length, decoding_scope,
                         output_fn, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param sequence_length: Sequence Length
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_fn: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: Train Logits
    """
    # TODO: Implement Function
    # Training Decoder
    train_decoder_fn = tf.contrib.seq2seq.simple_decoder_fn_train(encoder_state)
    train_pred, _, _ = tf.contrib.seq2seq.dynamic_rnn_decoder(
    dec_cell, train_decoder_fn, dec_embed_input, sequence_length, scope=decoding_scope)
    
    # Apply output function
    train_logits =  output_fn(train_pred)

    return train_logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### 解码 - 推论

使用 [`tf.contrib.seq2seq.simple_decoder_fn_inference()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/simple_decoder_fn_inference) 和 [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder) 创建推论分对数（inference logits）。

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id,
                         maximum_length, vocab_size, decoding_scope, output_fn, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param maximum_length: The maximum allowed time steps to decode
    :param vocab_size: Size of vocabulary
    :param decoding_scope: TensorFlow Variable Scope for decoding
    :param output_fn: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: Inference Logits
    """
    # TODO: Implement Function
    # Inference Decoder
    infer_decoder_fn = tf.contrib.seq2seq.simple_decoder_fn_inference(
        output_fn, encoder_state, dec_embeddings, start_of_sequence_id, end_of_sequence_id, 
        maximum_length - 1, vocab_size)
    inference_logits, _, _ = tf.contrib.seq2seq.dynamic_rnn_decoder(dec_cell, infer_decoder_fn, scope=decoding_scope)

    return inference_logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### 构建解码层级

实现 `decoding_layer()` 以创建解码器 RNN 层级。

- 使用 `rnn_size` 和 `num_layers` 创建解码 RNN 单元。
- 使用 [`lambda`](https://docs.python.org/3/tutorial/controlflow.html#lambda-expressions) 创建输出函数，将输入，也就是分对数转换为类分对数（class logits）。
- 使用 `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, sequence_length, decoding_scope, output_fn, keep_prob)` 函数获取训练分对数。
- 使用 `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, maximum_length, vocab_size, decoding_scope, output_fn, keep_prob)` 函数获取推论分对数。

注意：你将需要使用 [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) 在训练和推论分对数间分享变量。

In [12]:
def decoding_layer(dec_embed_input, dec_embeddings, encoder_state, vocab_size, sequence_length, rnn_size,
                   num_layers, target_vocab_to_int, keep_prob):
    """
    Create decoding layer
    :param dec_embed_input: Decoder embedded input
    :param dec_embeddings: Decoder embeddings
    :param encoder_state: The encoded state
    :param vocab_size: Size of vocabulary
    :param sequence_length: Sequence Length
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param keep_prob: Dropout keep probability
    :return: Tuple of (Training Logits, Inference Logits)
    """
    # TODO: Implement Function
    lstm = tf.contrib.rnn.BasicLSTMCell(rnn_size)
    dropout = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
    dec_cell = tf.contrib.rnn.MultiRNNCell([dropout] * num_layers)
    
    output_fn = lambda x: tf.contrib.layers.fully_connected(x, vocab_size, activation_fn=None, scope=decoding_scope)           

    with tf.variable_scope("decoding") as decoding_scope:
        training_decoder_output = decoding_layer_train(encoder_state, dec_cell, dec_embed_input, sequence_length, decoding_scope,
                         output_fn, keep_prob)

    with tf.variable_scope("decoding", reuse=True) as decoding_scope:
        start_of_sequence_id = target_vocab_to_int["<GO>"]
        end_of_sequence_id = target_vocab_to_int["<EOS>"]
        inference_decoder_output = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, 
                         sequence_length, vocab_size, decoding_scope, output_fn, keep_prob)
    
    return training_decoder_output, inference_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### 构建神经网络

应用你在上方实现的函数，以：

- 向编码器的输入数据应用嵌入。
- 使用 `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob)` 编码输入。
- 使用 `process_decoding_input(target_data, target_vocab_to_int, batch_size)` 函数处理目标数据。
- 向解码器的目标数据应用嵌入。
- 使用 `decoding_layer(dec_embed_input, dec_embeddings, encoder_state, vocab_size, sequence_length, rnn_size, num_layers, target_vocab_to_int, keep_prob)` 解码编码的输入数据。

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size, sequence_length, source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size, rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param sequence_length: Sequence Length
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training Logits, Inference Logits)
    """
    # TODO: Implement Function
    rnn_inputs = tf.contrib.layers.embed_sequence(input_data, source_vocab_size, enc_embedding_size)
    
    encoder_state = encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob)
    
    dec_input = process_decoding_input(target_data, target_vocab_to_int, batch_size)
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, dec_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
    
    train_logits, inference_logits = decoding_layer(dec_embed_input, dec_embeddings, encoder_state, target_vocab_size, sequence_length,
                       rnn_size, num_layers, target_vocab_to_int, keep_prob)
    
    return train_logits, inference_logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## 训练神经网络

### 超参数

调试以下参数：

- 将 `epochs` 设为 epoch 次数。
- 将 `batch_size` 设为批次大小。
- 将 `rnn_size` 设为 RNN 的大小。
- 将 `num_layers` 设为层级数量。
- 将 `encoding_embedding_size` 设为编码器嵌入大小。
- 将 `decoding_embedding_size` 设为解码器嵌入大小
- 将 `learning_rate` 设为训练速率。
- 将 `keep_probability` 设为丢弃保留率（Dropout keep probability）。

In [14]:
# Number of Epochs
epochs = 5
# Batch Size
batch_size = 128
# RNN Size
rnn_size = 128
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 100
decoding_embedding_size = 100
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.8

### 构建图表

使用你实现的神经网络构建图表。

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_source_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob = model_inputs()
    sequence_length = tf.placeholder_with_default(max_source_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)
    
    train_logits, inference_logits = seq2seq_model(
        tf.reverse(input_data, [-1]), targets, keep_prob, batch_size, sequence_length, len(source_vocab_to_int), len(target_vocab_to_int),
        encoding_embedding_size, decoding_embedding_size, rnn_size, num_layers, target_vocab_to_int)

    tf.identity(inference_logits, 'logits')
    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            train_logits,
            targets,
            tf.ones([input_shape[0], sequence_length]))

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)

### 训练

利用预处理的数据训练神经网络。如果很难获得低损失值，请访问我们的论坛，看看其他人是否遇到了相同的问题。

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import time

def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1]), (0,0)],
            'constant')

    return np.mean(np.equal(target, np.argmax(logits, 2)))

train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]

valid_source = helper.pad_sentence_batch(source_int_text[:batch_size])
valid_target = helper.pad_sentence_batch(target_int_text[:batch_size])

with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch) in enumerate(
                helper.batch_data(train_source, train_target, batch_size)):
            start_time = time.time()
            
            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 sequence_length: target_batch.shape[1],
                 keep_prob: keep_probability})
            
            batch_train_logits = sess.run(
                inference_logits,
                {input_data: source_batch, keep_prob: 1.0})
            batch_valid_logits = sess.run(
                inference_logits,
                {input_data: valid_source, keep_prob: 1.0})
                
            train_acc = get_accuracy(target_batch, batch_train_logits)
            valid_acc = get_accuracy(np.array(valid_target), batch_valid_logits)
            end_time = time.time()
            print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.3f}, Validation Accuracy: {:>6.3f}, Loss: {:>6.3f}'
                  .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    0/1077 - Train Accuracy:  0.294, Validation Accuracy:  0.305, Loss:  5.889
Epoch   0 Batch    1/1077 - Train Accuracy:  0.221, Validation Accuracy:  0.305, Loss:  5.074
Epoch   0 Batch    2/1077 - Train Accuracy:  0.244, Validation Accuracy:  0.335, Loss:  4.336
Epoch   0 Batch    3/1077 - Train Accuracy:  0.275, Validation Accuracy:  0.337, Loss:  3.832
Epoch   0 Batch    4/1077 - Train Accuracy:  0.263, Validation Accuracy:  0.336, Loss:  3.669
Epoch   0 Batch    5/1077 - Train Accuracy:  0.295, Validation Accuracy:  0.336, Loss:  3.552
Epoch   0 Batch    6/1077 - Train Accuracy:  0.282, Validation Accuracy:  0.342, Loss:  3.504
Epoch   0 Batch    7/1077 - Train Accuracy:  0.268, Validation Accuracy:  0.342, Loss:  3.515
Epoch   0 Batch    8/1077 - Train Accuracy:  0.275, Validation Accuracy:  0.341, Loss:  3.414
Epoch   0 Batch    9/1077 - Train Accuracy:  0.283, Validation Accuracy:  0.341, Loss:  3.324
Epoch   0 Batch   10/1077 - Train Accuracy:  0.273, Validati

Epoch   0 Batch   88/1077 - Train Accuracy:  0.509, Validation Accuracy:  0.533, Loss:  1.659
Epoch   0 Batch   89/1077 - Train Accuracy:  0.499, Validation Accuracy:  0.548, Loss:  1.632
Epoch   0 Batch   90/1077 - Train Accuracy:  0.471, Validation Accuracy:  0.535, Loss:  1.699
Epoch   0 Batch   91/1077 - Train Accuracy:  0.511, Validation Accuracy:  0.518, Loss:  1.448
Epoch   0 Batch   92/1077 - Train Accuracy:  0.490, Validation Accuracy:  0.524, Loss:  1.569
Epoch   0 Batch   93/1077 - Train Accuracy:  0.476, Validation Accuracy:  0.520, Loss:  1.581
Epoch   0 Batch   94/1077 - Train Accuracy:  0.485, Validation Accuracy:  0.526, Loss:  1.457
Epoch   0 Batch   95/1077 - Train Accuracy:  0.525, Validation Accuracy:  0.550, Loss:  1.486
Epoch   0 Batch   96/1077 - Train Accuracy:  0.477, Validation Accuracy:  0.533, Loss:  1.471
Epoch   0 Batch   97/1077 - Train Accuracy:  0.467, Validation Accuracy:  0.525, Loss:  1.476
Epoch   0 Batch   98/1077 - Train Accuracy:  0.497, Validati

Epoch   0 Batch  176/1077 - Train Accuracy:  0.565, Validation Accuracy:  0.605, Loss:  0.783
Epoch   0 Batch  177/1077 - Train Accuracy:  0.537, Validation Accuracy:  0.591, Loss:  0.837
Epoch   0 Batch  178/1077 - Train Accuracy:  0.564, Validation Accuracy:  0.582, Loss:  0.763
Epoch   0 Batch  179/1077 - Train Accuracy:  0.586, Validation Accuracy:  0.578, Loss:  0.807
Epoch   0 Batch  180/1077 - Train Accuracy:  0.559, Validation Accuracy:  0.581, Loss:  0.773
Epoch   0 Batch  181/1077 - Train Accuracy:  0.555, Validation Accuracy:  0.580, Loss:  0.803
Epoch   0 Batch  182/1077 - Train Accuracy:  0.596, Validation Accuracy:  0.570, Loss:  0.752
Epoch   0 Batch  183/1077 - Train Accuracy:  0.567, Validation Accuracy:  0.559, Loss:  0.767
Epoch   0 Batch  184/1077 - Train Accuracy:  0.568, Validation Accuracy:  0.566, Loss:  0.718
Epoch   0 Batch  185/1077 - Train Accuracy:  0.580, Validation Accuracy:  0.580, Loss:  0.760
Epoch   0 Batch  186/1077 - Train Accuracy:  0.569, Validati

Epoch   0 Batch  264/1077 - Train Accuracy:  0.626, Validation Accuracy:  0.630, Loss:  0.619
Epoch   0 Batch  265/1077 - Train Accuracy:  0.617, Validation Accuracy:  0.625, Loss:  0.626
Epoch   0 Batch  266/1077 - Train Accuracy:  0.638, Validation Accuracy:  0.617, Loss:  0.590
Epoch   0 Batch  267/1077 - Train Accuracy:  0.656, Validation Accuracy:  0.608, Loss:  0.576
Epoch   0 Batch  268/1077 - Train Accuracy:  0.650, Validation Accuracy:  0.623, Loss:  0.614
Epoch   0 Batch  269/1077 - Train Accuracy:  0.624, Validation Accuracy:  0.627, Loss:  0.658
Epoch   0 Batch  270/1077 - Train Accuracy:  0.610, Validation Accuracy:  0.622, Loss:  0.650
Epoch   0 Batch  271/1077 - Train Accuracy:  0.672, Validation Accuracy:  0.624, Loss:  0.607
Epoch   0 Batch  272/1077 - Train Accuracy:  0.635, Validation Accuracy:  0.630, Loss:  0.633
Epoch   0 Batch  273/1077 - Train Accuracy:  0.629, Validation Accuracy:  0.633, Loss:  0.586
Epoch   0 Batch  274/1077 - Train Accuracy:  0.664, Validati

Epoch   0 Batch  352/1077 - Train Accuracy:  0.664, Validation Accuracy:  0.655, Loss:  0.523
Epoch   0 Batch  353/1077 - Train Accuracy:  0.635, Validation Accuracy:  0.670, Loss:  0.567
Epoch   0 Batch  354/1077 - Train Accuracy:  0.650, Validation Accuracy:  0.661, Loss:  0.548
Epoch   0 Batch  355/1077 - Train Accuracy:  0.652, Validation Accuracy:  0.649, Loss:  0.516
Epoch   0 Batch  356/1077 - Train Accuracy:  0.686, Validation Accuracy:  0.643, Loss:  0.528
Epoch   0 Batch  357/1077 - Train Accuracy:  0.671, Validation Accuracy:  0.644, Loss:  0.501
Epoch   0 Batch  358/1077 - Train Accuracy:  0.636, Validation Accuracy:  0.648, Loss:  0.546
Epoch   0 Batch  359/1077 - Train Accuracy:  0.663, Validation Accuracy:  0.662, Loss:  0.524
Epoch   0 Batch  360/1077 - Train Accuracy:  0.651, Validation Accuracy:  0.648, Loss:  0.515
Epoch   0 Batch  361/1077 - Train Accuracy:  0.680, Validation Accuracy:  0.657, Loss:  0.545
Epoch   0 Batch  362/1077 - Train Accuracy:  0.670, Validati

Epoch   0 Batch  440/1077 - Train Accuracy:  0.701, Validation Accuracy:  0.705, Loss:  0.441
Epoch   0 Batch  441/1077 - Train Accuracy:  0.691, Validation Accuracy:  0.699, Loss:  0.414
Epoch   0 Batch  442/1077 - Train Accuracy:  0.730, Validation Accuracy:  0.701, Loss:  0.412
Epoch   0 Batch  443/1077 - Train Accuracy:  0.728, Validation Accuracy:  0.702, Loss:  0.403
Epoch   0 Batch  444/1077 - Train Accuracy:  0.741, Validation Accuracy:  0.677, Loss:  0.426
Epoch   0 Batch  445/1077 - Train Accuracy:  0.686, Validation Accuracy:  0.675, Loss:  0.439
Epoch   0 Batch  446/1077 - Train Accuracy:  0.737, Validation Accuracy:  0.666, Loss:  0.391
Epoch   0 Batch  447/1077 - Train Accuracy:  0.677, Validation Accuracy:  0.670, Loss:  0.408
Epoch   0 Batch  448/1077 - Train Accuracy:  0.691, Validation Accuracy:  0.669, Loss:  0.430
Epoch   0 Batch  449/1077 - Train Accuracy:  0.672, Validation Accuracy:  0.686, Loss:  0.430
Epoch   0 Batch  450/1077 - Train Accuracy:  0.721, Validati

Epoch   0 Batch  528/1077 - Train Accuracy:  0.746, Validation Accuracy:  0.724, Loss:  0.349
Epoch   0 Batch  529/1077 - Train Accuracy:  0.725, Validation Accuracy:  0.728, Loss:  0.349
Epoch   0 Batch  530/1077 - Train Accuracy:  0.761, Validation Accuracy:  0.752, Loss:  0.358
Epoch   0 Batch  531/1077 - Train Accuracy:  0.751, Validation Accuracy:  0.731, Loss:  0.346
Epoch   0 Batch  532/1077 - Train Accuracy:  0.717, Validation Accuracy:  0.741, Loss:  0.376
Epoch   0 Batch  533/1077 - Train Accuracy:  0.760, Validation Accuracy:  0.744, Loss:  0.347
Epoch   0 Batch  534/1077 - Train Accuracy:  0.766, Validation Accuracy:  0.729, Loss:  0.345
Epoch   0 Batch  535/1077 - Train Accuracy:  0.786, Validation Accuracy:  0.743, Loss:  0.329
Epoch   0 Batch  536/1077 - Train Accuracy:  0.740, Validation Accuracy:  0.745, Loss:  0.348
Epoch   0 Batch  537/1077 - Train Accuracy:  0.760, Validation Accuracy:  0.762, Loss:  0.339
Epoch   0 Batch  538/1077 - Train Accuracy:  0.788, Validati

Epoch   0 Batch  616/1077 - Train Accuracy:  0.792, Validation Accuracy:  0.767, Loss:  0.284
Epoch   0 Batch  617/1077 - Train Accuracy:  0.806, Validation Accuracy:  0.771, Loss:  0.262
Epoch   0 Batch  618/1077 - Train Accuracy:  0.802, Validation Accuracy:  0.778, Loss:  0.270
Epoch   0 Batch  619/1077 - Train Accuracy:  0.789, Validation Accuracy:  0.779, Loss:  0.266
Epoch   0 Batch  620/1077 - Train Accuracy:  0.828, Validation Accuracy:  0.777, Loss:  0.254
Epoch   0 Batch  621/1077 - Train Accuracy:  0.854, Validation Accuracy:  0.780, Loss:  0.259
Epoch   0 Batch  622/1077 - Train Accuracy:  0.834, Validation Accuracy:  0.774, Loss:  0.280
Epoch   0 Batch  623/1077 - Train Accuracy:  0.787, Validation Accuracy:  0.766, Loss:  0.270
Epoch   0 Batch  624/1077 - Train Accuracy:  0.828, Validation Accuracy:  0.771, Loss:  0.253
Epoch   0 Batch  625/1077 - Train Accuracy:  0.818, Validation Accuracy:  0.782, Loss:  0.259
Epoch   0 Batch  626/1077 - Train Accuracy:  0.804, Validati

Epoch   0 Batch  704/1077 - Train Accuracy:  0.845, Validation Accuracy:  0.820, Loss:  0.235
Epoch   0 Batch  705/1077 - Train Accuracy:  0.846, Validation Accuracy:  0.819, Loss:  0.250
Epoch   0 Batch  706/1077 - Train Accuracy:  0.806, Validation Accuracy:  0.828, Loss:  0.244
Epoch   0 Batch  707/1077 - Train Accuracy:  0.854, Validation Accuracy:  0.832, Loss:  0.211
Epoch   0 Batch  708/1077 - Train Accuracy:  0.821, Validation Accuracy:  0.837, Loss:  0.216
Epoch   0 Batch  709/1077 - Train Accuracy:  0.834, Validation Accuracy:  0.838, Loss:  0.238
Epoch   0 Batch  710/1077 - Train Accuracy:  0.851, Validation Accuracy:  0.836, Loss:  0.197
Epoch   0 Batch  711/1077 - Train Accuracy:  0.820, Validation Accuracy:  0.838, Loss:  0.229
Epoch   0 Batch  712/1077 - Train Accuracy:  0.854, Validation Accuracy:  0.845, Loss:  0.196
Epoch   0 Batch  713/1077 - Train Accuracy:  0.839, Validation Accuracy:  0.825, Loss:  0.172
Epoch   0 Batch  714/1077 - Train Accuracy:  0.829, Validati

Epoch   0 Batch  792/1077 - Train Accuracy:  0.855, Validation Accuracy:  0.869, Loss:  0.185
Epoch   0 Batch  793/1077 - Train Accuracy:  0.901, Validation Accuracy:  0.871, Loss:  0.172
Epoch   0 Batch  794/1077 - Train Accuracy:  0.866, Validation Accuracy:  0.881, Loss:  0.152
Epoch   0 Batch  795/1077 - Train Accuracy:  0.843, Validation Accuracy:  0.876, Loss:  0.178
Epoch   0 Batch  796/1077 - Train Accuracy:  0.880, Validation Accuracy:  0.874, Loss:  0.168
Epoch   0 Batch  797/1077 - Train Accuracy:  0.879, Validation Accuracy:  0.864, Loss:  0.159
Epoch   0 Batch  798/1077 - Train Accuracy:  0.863, Validation Accuracy:  0.857, Loss:  0.175
Epoch   0 Batch  799/1077 - Train Accuracy:  0.821, Validation Accuracy:  0.853, Loss:  0.194
Epoch   0 Batch  800/1077 - Train Accuracy:  0.859, Validation Accuracy:  0.863, Loss:  0.172
Epoch   0 Batch  801/1077 - Train Accuracy:  0.871, Validation Accuracy:  0.851, Loss:  0.168
Epoch   0 Batch  802/1077 - Train Accuracy:  0.834, Validati

Epoch   0 Batch  880/1077 - Train Accuracy:  0.877, Validation Accuracy:  0.868, Loss:  0.152
Epoch   0 Batch  881/1077 - Train Accuracy:  0.867, Validation Accuracy:  0.864, Loss:  0.146
Epoch   0 Batch  882/1077 - Train Accuracy:  0.874, Validation Accuracy:  0.868, Loss:  0.163
Epoch   0 Batch  883/1077 - Train Accuracy:  0.873, Validation Accuracy:  0.859, Loss:  0.183
Epoch   0 Batch  884/1077 - Train Accuracy:  0.867, Validation Accuracy:  0.865, Loss:  0.131
Epoch   0 Batch  885/1077 - Train Accuracy:  0.873, Validation Accuracy:  0.869, Loss:  0.118
Epoch   0 Batch  886/1077 - Train Accuracy:  0.877, Validation Accuracy:  0.861, Loss:  0.139
Epoch   0 Batch  887/1077 - Train Accuracy:  0.857, Validation Accuracy:  0.861, Loss:  0.168
Epoch   0 Batch  888/1077 - Train Accuracy:  0.900, Validation Accuracy:  0.865, Loss:  0.134
Epoch   0 Batch  889/1077 - Train Accuracy:  0.880, Validation Accuracy:  0.858, Loss:  0.139
Epoch   0 Batch  890/1077 - Train Accuracy:  0.903, Validati

Epoch   0 Batch  968/1077 - Train Accuracy:  0.846, Validation Accuracy:  0.881, Loss:  0.140
Epoch   0 Batch  969/1077 - Train Accuracy:  0.861, Validation Accuracy:  0.885, Loss:  0.133
Epoch   0 Batch  970/1077 - Train Accuracy:  0.902, Validation Accuracy:  0.886, Loss:  0.128
Epoch   0 Batch  971/1077 - Train Accuracy:  0.903, Validation Accuracy:  0.886, Loss:  0.122
Epoch   0 Batch  972/1077 - Train Accuracy:  0.883, Validation Accuracy:  0.883, Loss:  0.114
Epoch   0 Batch  973/1077 - Train Accuracy:  0.916, Validation Accuracy:  0.877, Loss:  0.098
Epoch   0 Batch  974/1077 - Train Accuracy:  0.889, Validation Accuracy:  0.873, Loss:  0.105
Epoch   0 Batch  975/1077 - Train Accuracy:  0.896, Validation Accuracy:  0.876, Loss:  0.112
Epoch   0 Batch  976/1077 - Train Accuracy:  0.898, Validation Accuracy:  0.877, Loss:  0.105
Epoch   0 Batch  977/1077 - Train Accuracy:  0.900, Validation Accuracy:  0.884, Loss:  0.105
Epoch   0 Batch  978/1077 - Train Accuracy:  0.894, Validati

Epoch   0 Batch 1056/1077 - Train Accuracy:  0.894, Validation Accuracy:  0.892, Loss:  0.099
Epoch   0 Batch 1057/1077 - Train Accuracy:  0.873, Validation Accuracy:  0.886, Loss:  0.119
Epoch   0 Batch 1058/1077 - Train Accuracy:  0.895, Validation Accuracy:  0.883, Loss:  0.116
Epoch   0 Batch 1059/1077 - Train Accuracy:  0.847, Validation Accuracy:  0.876, Loss:  0.138
Epoch   0 Batch 1060/1077 - Train Accuracy:  0.903, Validation Accuracy:  0.876, Loss:  0.100
Epoch   0 Batch 1061/1077 - Train Accuracy:  0.875, Validation Accuracy:  0.880, Loss:  0.123
Epoch   0 Batch 1062/1077 - Train Accuracy:  0.872, Validation Accuracy:  0.879, Loss:  0.107
Epoch   0 Batch 1063/1077 - Train Accuracy:  0.879, Validation Accuracy:  0.885, Loss:  0.107
Epoch   0 Batch 1064/1077 - Train Accuracy:  0.920, Validation Accuracy:  0.888, Loss:  0.107
Epoch   0 Batch 1065/1077 - Train Accuracy:  0.907, Validation Accuracy:  0.889, Loss:  0.096
Epoch   0 Batch 1066/1077 - Train Accuracy:  0.907, Validati

Epoch   1 Batch   68/1077 - Train Accuracy:  0.891, Validation Accuracy:  0.906, Loss:  0.092
Epoch   1 Batch   69/1077 - Train Accuracy:  0.902, Validation Accuracy:  0.912, Loss:  0.109
Epoch   1 Batch   70/1077 - Train Accuracy:  0.907, Validation Accuracy:  0.906, Loss:  0.094
Epoch   1 Batch   71/1077 - Train Accuracy:  0.928, Validation Accuracy:  0.901, Loss:  0.064
Epoch   1 Batch   72/1077 - Train Accuracy:  0.875, Validation Accuracy:  0.901, Loss:  0.097
Epoch   1 Batch   73/1077 - Train Accuracy:  0.918, Validation Accuracy:  0.901, Loss:  0.093
Epoch   1 Batch   74/1077 - Train Accuracy:  0.892, Validation Accuracy:  0.898, Loss:  0.082
Epoch   1 Batch   75/1077 - Train Accuracy:  0.895, Validation Accuracy:  0.902, Loss:  0.122
Epoch   1 Batch   76/1077 - Train Accuracy:  0.906, Validation Accuracy:  0.892, Loss:  0.076
Epoch   1 Batch   77/1077 - Train Accuracy:  0.895, Validation Accuracy:  0.892, Loss:  0.085
Epoch   1 Batch   78/1077 - Train Accuracy:  0.897, Validati

Epoch   1 Batch  156/1077 - Train Accuracy:  0.898, Validation Accuracy:  0.887, Loss:  0.079
Epoch   1 Batch  157/1077 - Train Accuracy:  0.909, Validation Accuracy:  0.894, Loss:  0.086
Epoch   1 Batch  158/1077 - Train Accuracy:  0.882, Validation Accuracy:  0.888, Loss:  0.111
Epoch   1 Batch  159/1077 - Train Accuracy:  0.929, Validation Accuracy:  0.892, Loss:  0.071
Epoch   1 Batch  160/1077 - Train Accuracy:  0.896, Validation Accuracy:  0.892, Loss:  0.076
Epoch   1 Batch  161/1077 - Train Accuracy:  0.886, Validation Accuracy:  0.893, Loss:  0.081
Epoch   1 Batch  162/1077 - Train Accuracy:  0.904, Validation Accuracy:  0.896, Loss:  0.103
Epoch   1 Batch  163/1077 - Train Accuracy:  0.905, Validation Accuracy:  0.893, Loss:  0.106
Epoch   1 Batch  164/1077 - Train Accuracy:  0.898, Validation Accuracy:  0.893, Loss:  0.087
Epoch   1 Batch  165/1077 - Train Accuracy:  0.916, Validation Accuracy:  0.903, Loss:  0.065
Epoch   1 Batch  166/1077 - Train Accuracy:  0.901, Validati

Epoch   1 Batch  244/1077 - Train Accuracy:  0.915, Validation Accuracy:  0.889, Loss:  0.076
Epoch   1 Batch  245/1077 - Train Accuracy:  0.917, Validation Accuracy:  0.907, Loss:  0.077
Epoch   1 Batch  246/1077 - Train Accuracy:  0.914, Validation Accuracy:  0.903, Loss:  0.072
Epoch   1 Batch  247/1077 - Train Accuracy:  0.922, Validation Accuracy:  0.905, Loss:  0.076
Epoch   1 Batch  248/1077 - Train Accuracy:  0.907, Validation Accuracy:  0.905, Loss:  0.090
Epoch   1 Batch  249/1077 - Train Accuracy:  0.918, Validation Accuracy:  0.909, Loss:  0.080
Epoch   1 Batch  250/1077 - Train Accuracy:  0.925, Validation Accuracy:  0.903, Loss:  0.073
Epoch   1 Batch  251/1077 - Train Accuracy:  0.906, Validation Accuracy:  0.902, Loss:  0.090
Epoch   1 Batch  252/1077 - Train Accuracy:  0.907, Validation Accuracy:  0.909, Loss:  0.081
Epoch   1 Batch  253/1077 - Train Accuracy:  0.916, Validation Accuracy:  0.908, Loss:  0.080
Epoch   1 Batch  254/1077 - Train Accuracy:  0.894, Validati

Epoch   1 Batch  332/1077 - Train Accuracy:  0.913, Validation Accuracy:  0.878, Loss:  0.066
Epoch   1 Batch  333/1077 - Train Accuracy:  0.926, Validation Accuracy:  0.882, Loss:  0.063
Epoch   1 Batch  334/1077 - Train Accuracy:  0.939, Validation Accuracy:  0.885, Loss:  0.074
Epoch   1 Batch  335/1077 - Train Accuracy:  0.920, Validation Accuracy:  0.888, Loss:  0.077
Epoch   1 Batch  336/1077 - Train Accuracy:  0.893, Validation Accuracy:  0.884, Loss:  0.108
Epoch   1 Batch  337/1077 - Train Accuracy:  0.895, Validation Accuracy:  0.880, Loss:  0.081
Epoch   1 Batch  338/1077 - Train Accuracy:  0.869, Validation Accuracy:  0.881, Loss:  0.106
Epoch   1 Batch  339/1077 - Train Accuracy:  0.922, Validation Accuracy:  0.881, Loss:  0.070
Epoch   1 Batch  340/1077 - Train Accuracy:  0.934, Validation Accuracy:  0.896, Loss:  0.070
Epoch   1 Batch  341/1077 - Train Accuracy:  0.898, Validation Accuracy:  0.901, Loss:  0.093
Epoch   1 Batch  342/1077 - Train Accuracy:  0.887, Validati

Epoch   1 Batch  420/1077 - Train Accuracy:  0.939, Validation Accuracy:  0.916, Loss:  0.059
Epoch   1 Batch  421/1077 - Train Accuracy:  0.895, Validation Accuracy:  0.915, Loss:  0.083
Epoch   1 Batch  422/1077 - Train Accuracy:  0.898, Validation Accuracy:  0.912, Loss:  0.069
Epoch   1 Batch  423/1077 - Train Accuracy:  0.909, Validation Accuracy:  0.910, Loss:  0.086
Epoch   1 Batch  424/1077 - Train Accuracy:  0.892, Validation Accuracy:  0.902, Loss:  0.073
Epoch   1 Batch  425/1077 - Train Accuracy:  0.917, Validation Accuracy:  0.900, Loss:  0.064
Epoch   1 Batch  426/1077 - Train Accuracy:  0.892, Validation Accuracy:  0.901, Loss:  0.089
Epoch   1 Batch  427/1077 - Train Accuracy:  0.886, Validation Accuracy:  0.905, Loss:  0.070
Epoch   1 Batch  428/1077 - Train Accuracy:  0.956, Validation Accuracy:  0.906, Loss:  0.059
Epoch   1 Batch  429/1077 - Train Accuracy:  0.924, Validation Accuracy:  0.902, Loss:  0.067
Epoch   1 Batch  430/1077 - Train Accuracy:  0.905, Validati

Epoch   1 Batch  508/1077 - Train Accuracy:  0.920, Validation Accuracy:  0.916, Loss:  0.059
Epoch   1 Batch  509/1077 - Train Accuracy:  0.886, Validation Accuracy:  0.914, Loss:  0.079
Epoch   1 Batch  510/1077 - Train Accuracy:  0.932, Validation Accuracy:  0.901, Loss:  0.072
Epoch   1 Batch  511/1077 - Train Accuracy:  0.903, Validation Accuracy:  0.906, Loss:  0.071
Epoch   1 Batch  512/1077 - Train Accuracy:  0.926, Validation Accuracy:  0.909, Loss:  0.055
Epoch   1 Batch  513/1077 - Train Accuracy:  0.901, Validation Accuracy:  0.900, Loss:  0.068
Epoch   1 Batch  514/1077 - Train Accuracy:  0.898, Validation Accuracy:  0.904, Loss:  0.080
Epoch   1 Batch  515/1077 - Train Accuracy:  0.925, Validation Accuracy:  0.910, Loss:  0.079
Epoch   1 Batch  516/1077 - Train Accuracy:  0.934, Validation Accuracy:  0.911, Loss:  0.072
Epoch   1 Batch  517/1077 - Train Accuracy:  0.918, Validation Accuracy:  0.911, Loss:  0.068
Epoch   1 Batch  518/1077 - Train Accuracy:  0.941, Validati

Epoch   1 Batch  596/1077 - Train Accuracy:  0.931, Validation Accuracy:  0.875, Loss:  0.070
Epoch   1 Batch  597/1077 - Train Accuracy:  0.894, Validation Accuracy:  0.893, Loss:  0.059
Epoch   1 Batch  598/1077 - Train Accuracy:  0.892, Validation Accuracy:  0.907, Loss:  0.080
Epoch   1 Batch  599/1077 - Train Accuracy:  0.895, Validation Accuracy:  0.907, Loss:  0.089
Epoch   1 Batch  600/1077 - Train Accuracy:  0.932, Validation Accuracy:  0.900, Loss:  0.078
Epoch   1 Batch  601/1077 - Train Accuracy:  0.900, Validation Accuracy:  0.902, Loss:  0.074
Epoch   1 Batch  602/1077 - Train Accuracy:  0.919, Validation Accuracy:  0.905, Loss:  0.070
Epoch   1 Batch  603/1077 - Train Accuracy:  0.910, Validation Accuracy:  0.910, Loss:  0.067
Epoch   1 Batch  604/1077 - Train Accuracy:  0.900, Validation Accuracy:  0.912, Loss:  0.079
Epoch   1 Batch  605/1077 - Train Accuracy:  0.902, Validation Accuracy:  0.905, Loss:  0.089
Epoch   1 Batch  606/1077 - Train Accuracy:  0.926, Validati

Epoch   1 Batch  684/1077 - Train Accuracy:  0.926, Validation Accuracy:  0.909, Loss:  0.067
Epoch   1 Batch  685/1077 - Train Accuracy:  0.897, Validation Accuracy:  0.914, Loss:  0.075
Epoch   1 Batch  686/1077 - Train Accuracy:  0.918, Validation Accuracy:  0.917, Loss:  0.060
Epoch   1 Batch  687/1077 - Train Accuracy:  0.921, Validation Accuracy:  0.913, Loss:  0.091
Epoch   1 Batch  688/1077 - Train Accuracy:  0.943, Validation Accuracy:  0.914, Loss:  0.066
Epoch   1 Batch  689/1077 - Train Accuracy:  0.939, Validation Accuracy:  0.904, Loss:  0.047
Epoch   1 Batch  690/1077 - Train Accuracy:  0.921, Validation Accuracy:  0.909, Loss:  0.068
Epoch   1 Batch  691/1077 - Train Accuracy:  0.899, Validation Accuracy:  0.906, Loss:  0.085
Epoch   1 Batch  692/1077 - Train Accuracy:  0.915, Validation Accuracy:  0.906, Loss:  0.061
Epoch   1 Batch  693/1077 - Train Accuracy:  0.879, Validation Accuracy:  0.901, Loss:  0.084
Epoch   1 Batch  694/1077 - Train Accuracy:  0.933, Validati

Epoch   1 Batch  772/1077 - Train Accuracy:  0.921, Validation Accuracy:  0.909, Loss:  0.060
Epoch   1 Batch  773/1077 - Train Accuracy:  0.896, Validation Accuracy:  0.909, Loss:  0.068
Epoch   1 Batch  774/1077 - Train Accuracy:  0.929, Validation Accuracy:  0.909, Loss:  0.067
Epoch   1 Batch  775/1077 - Train Accuracy:  0.923, Validation Accuracy:  0.907, Loss:  0.072
Epoch   1 Batch  776/1077 - Train Accuracy:  0.916, Validation Accuracy:  0.924, Loss:  0.058
Epoch   1 Batch  777/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.926, Loss:  0.070
Epoch   1 Batch  778/1077 - Train Accuracy:  0.906, Validation Accuracy:  0.924, Loss:  0.064
Epoch   1 Batch  779/1077 - Train Accuracy:  0.908, Validation Accuracy:  0.920, Loss:  0.069
Epoch   1 Batch  780/1077 - Train Accuracy:  0.889, Validation Accuracy:  0.916, Loss:  0.091
Epoch   1 Batch  781/1077 - Train Accuracy:  0.936, Validation Accuracy:  0.923, Loss:  0.054
Epoch   1 Batch  782/1077 - Train Accuracy:  0.914, Validati

Epoch   1 Batch  860/1077 - Train Accuracy:  0.913, Validation Accuracy:  0.900, Loss:  0.081
Epoch   1 Batch  861/1077 - Train Accuracy:  0.916, Validation Accuracy:  0.900, Loss:  0.067
Epoch   1 Batch  862/1077 - Train Accuracy:  0.923, Validation Accuracy:  0.901, Loss:  0.067
Epoch   1 Batch  863/1077 - Train Accuracy:  0.928, Validation Accuracy:  0.909, Loss:  0.054
Epoch   1 Batch  864/1077 - Train Accuracy:  0.899, Validation Accuracy:  0.898, Loss:  0.073
Epoch   1 Batch  865/1077 - Train Accuracy:  0.916, Validation Accuracy:  0.896, Loss:  0.069
Epoch   1 Batch  866/1077 - Train Accuracy:  0.905, Validation Accuracy:  0.913, Loss:  0.078
Epoch   1 Batch  867/1077 - Train Accuracy:  0.902, Validation Accuracy:  0.911, Loss:  0.115
Epoch   1 Batch  868/1077 - Train Accuracy:  0.914, Validation Accuracy:  0.909, Loss:  0.069
Epoch   1 Batch  869/1077 - Train Accuracy:  0.912, Validation Accuracy:  0.910, Loss:  0.067
Epoch   1 Batch  870/1077 - Train Accuracy:  0.912, Validati

Epoch   1 Batch  948/1077 - Train Accuracy:  0.911, Validation Accuracy:  0.909, Loss:  0.059
Epoch   1 Batch  949/1077 - Train Accuracy:  0.965, Validation Accuracy:  0.915, Loss:  0.052
Epoch   1 Batch  950/1077 - Train Accuracy:  0.942, Validation Accuracy:  0.917, Loss:  0.058
Epoch   1 Batch  951/1077 - Train Accuracy:  0.900, Validation Accuracy:  0.923, Loss:  0.068
Epoch   1 Batch  952/1077 - Train Accuracy:  0.932, Validation Accuracy:  0.928, Loss:  0.055
Epoch   1 Batch  953/1077 - Train Accuracy:  0.933, Validation Accuracy:  0.930, Loss:  0.049
Epoch   1 Batch  954/1077 - Train Accuracy:  0.912, Validation Accuracy:  0.925, Loss:  0.067
Epoch   1 Batch  955/1077 - Train Accuracy:  0.905, Validation Accuracy:  0.928, Loss:  0.074
Epoch   1 Batch  956/1077 - Train Accuracy:  0.905, Validation Accuracy:  0.933, Loss:  0.066
Epoch   1 Batch  957/1077 - Train Accuracy:  0.929, Validation Accuracy:  0.922, Loss:  0.050
Epoch   1 Batch  958/1077 - Train Accuracy:  0.932, Validati

Epoch   1 Batch 1036/1077 - Train Accuracy:  0.911, Validation Accuracy:  0.919, Loss:  0.076
Epoch   1 Batch 1037/1077 - Train Accuracy:  0.903, Validation Accuracy:  0.924, Loss:  0.061
Epoch   1 Batch 1038/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.923, Loss:  0.076
Epoch   1 Batch 1039/1077 - Train Accuracy:  0.922, Validation Accuracy:  0.929, Loss:  0.066
Epoch   1 Batch 1040/1077 - Train Accuracy:  0.924, Validation Accuracy:  0.928, Loss:  0.061
Epoch   1 Batch 1041/1077 - Train Accuracy:  0.915, Validation Accuracy:  0.928, Loss:  0.067
Epoch   1 Batch 1042/1077 - Train Accuracy:  0.928, Validation Accuracy:  0.920, Loss:  0.052
Epoch   1 Batch 1043/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.929, Loss:  0.076
Epoch   1 Batch 1044/1077 - Train Accuracy:  0.924, Validation Accuracy:  0.924, Loss:  0.081
Epoch   1 Batch 1045/1077 - Train Accuracy:  0.919, Validation Accuracy:  0.913, Loss:  0.058
Epoch   1 Batch 1046/1077 - Train Accuracy:  0.922, Validati

Epoch   2 Batch   48/1077 - Train Accuracy:  0.898, Validation Accuracy:  0.912, Loss:  0.084
Epoch   2 Batch   49/1077 - Train Accuracy:  0.926, Validation Accuracy:  0.904, Loss:  0.078
Epoch   2 Batch   50/1077 - Train Accuracy:  0.929, Validation Accuracy:  0.902, Loss:  0.058
Epoch   2 Batch   51/1077 - Train Accuracy:  0.908, Validation Accuracy:  0.906, Loss:  0.073
Epoch   2 Batch   52/1077 - Train Accuracy:  0.918, Validation Accuracy:  0.907, Loss:  0.065
Epoch   2 Batch   53/1077 - Train Accuracy:  0.923, Validation Accuracy:  0.915, Loss:  0.058
Epoch   2 Batch   54/1077 - Train Accuracy:  0.920, Validation Accuracy:  0.910, Loss:  0.083
Epoch   2 Batch   55/1077 - Train Accuracy:  0.945, Validation Accuracy:  0.911, Loss:  0.056
Epoch   2 Batch   56/1077 - Train Accuracy:  0.947, Validation Accuracy:  0.907, Loss:  0.044
Epoch   2 Batch   57/1077 - Train Accuracy:  0.911, Validation Accuracy:  0.909, Loss:  0.054
Epoch   2 Batch   58/1077 - Train Accuracy:  0.936, Validati

Epoch   2 Batch  136/1077 - Train Accuracy:  0.950, Validation Accuracy:  0.904, Loss:  0.050
Epoch   2 Batch  137/1077 - Train Accuracy:  0.943, Validation Accuracy:  0.904, Loss:  0.041
Epoch   2 Batch  138/1077 - Train Accuracy:  0.905, Validation Accuracy:  0.907, Loss:  0.054
Epoch   2 Batch  139/1077 - Train Accuracy:  0.921, Validation Accuracy:  0.903, Loss:  0.059
Epoch   2 Batch  140/1077 - Train Accuracy:  0.944, Validation Accuracy:  0.924, Loss:  0.053
Epoch   2 Batch  141/1077 - Train Accuracy:  0.938, Validation Accuracy:  0.926, Loss:  0.052
Epoch   2 Batch  142/1077 - Train Accuracy:  0.924, Validation Accuracy:  0.917, Loss:  0.052
Epoch   2 Batch  143/1077 - Train Accuracy:  0.902, Validation Accuracy:  0.925, Loss:  0.062
Epoch   2 Batch  144/1077 - Train Accuracy:  0.898, Validation Accuracy:  0.927, Loss:  0.087
Epoch   2 Batch  145/1077 - Train Accuracy:  0.935, Validation Accuracy:  0.923, Loss:  0.070
Epoch   2 Batch  146/1077 - Train Accuracy:  0.909, Validati

Epoch   2 Batch  224/1077 - Train Accuracy:  0.917, Validation Accuracy:  0.912, Loss:  0.064
Epoch   2 Batch  225/1077 - Train Accuracy:  0.932, Validation Accuracy:  0.917, Loss:  0.058
Epoch   2 Batch  226/1077 - Train Accuracy:  0.911, Validation Accuracy:  0.921, Loss:  0.050
Epoch   2 Batch  227/1077 - Train Accuracy:  0.892, Validation Accuracy:  0.911, Loss:  0.073
Epoch   2 Batch  228/1077 - Train Accuracy:  0.936, Validation Accuracy:  0.915, Loss:  0.053
Epoch   2 Batch  229/1077 - Train Accuracy:  0.914, Validation Accuracy:  0.930, Loss:  0.058
Epoch   2 Batch  230/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.914, Loss:  0.056
Epoch   2 Batch  231/1077 - Train Accuracy:  0.921, Validation Accuracy:  0.914, Loss:  0.050
Epoch   2 Batch  232/1077 - Train Accuracy:  0.925, Validation Accuracy:  0.921, Loss:  0.041
Epoch   2 Batch  233/1077 - Train Accuracy:  0.920, Validation Accuracy:  0.911, Loss:  0.075
Epoch   2 Batch  234/1077 - Train Accuracy:  0.919, Validati

Epoch   2 Batch  312/1077 - Train Accuracy:  0.920, Validation Accuracy:  0.904, Loss:  0.077
Epoch   2 Batch  313/1077 - Train Accuracy:  0.948, Validation Accuracy:  0.899, Loss:  0.049
Epoch   2 Batch  314/1077 - Train Accuracy:  0.945, Validation Accuracy:  0.903, Loss:  0.049
Epoch   2 Batch  315/1077 - Train Accuracy:  0.957, Validation Accuracy:  0.899, Loss:  0.049
Epoch   2 Batch  316/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.893, Loss:  0.068
Epoch   2 Batch  317/1077 - Train Accuracy:  0.924, Validation Accuracy:  0.895, Loss:  0.082
Epoch   2 Batch  318/1077 - Train Accuracy:  0.950, Validation Accuracy:  0.898, Loss:  0.066
Epoch   2 Batch  319/1077 - Train Accuracy:  0.933, Validation Accuracy:  0.900, Loss:  0.071
Epoch   2 Batch  320/1077 - Train Accuracy:  0.934, Validation Accuracy:  0.893, Loss:  0.067
Epoch   2 Batch  321/1077 - Train Accuracy:  0.915, Validation Accuracy:  0.898, Loss:  0.054
Epoch   2 Batch  322/1077 - Train Accuracy:  0.904, Validati

Epoch   2 Batch  400/1077 - Train Accuracy:  0.926, Validation Accuracy:  0.943, Loss:  0.071
Epoch   2 Batch  401/1077 - Train Accuracy:  0.945, Validation Accuracy:  0.935, Loss:  0.049
Epoch   2 Batch  402/1077 - Train Accuracy:  0.931, Validation Accuracy:  0.931, Loss:  0.050
Epoch   2 Batch  403/1077 - Train Accuracy:  0.884, Validation Accuracy:  0.927, Loss:  0.077
Epoch   2 Batch  404/1077 - Train Accuracy:  0.959, Validation Accuracy:  0.929, Loss:  0.050
Epoch   2 Batch  405/1077 - Train Accuracy:  0.921, Validation Accuracy:  0.930, Loss:  0.065
Epoch   2 Batch  406/1077 - Train Accuracy:  0.943, Validation Accuracy:  0.930, Loss:  0.051
Epoch   2 Batch  407/1077 - Train Accuracy:  0.918, Validation Accuracy:  0.922, Loss:  0.062
Epoch   2 Batch  408/1077 - Train Accuracy:  0.902, Validation Accuracy:  0.917, Loss:  0.052
Epoch   2 Batch  409/1077 - Train Accuracy:  0.912, Validation Accuracy:  0.917, Loss:  0.066
Epoch   2 Batch  410/1077 - Train Accuracy:  0.916, Validati

Epoch   2 Batch  488/1077 - Train Accuracy:  0.944, Validation Accuracy:  0.914, Loss:  0.050
Epoch   2 Batch  489/1077 - Train Accuracy:  0.915, Validation Accuracy:  0.909, Loss:  0.048
Epoch   2 Batch  490/1077 - Train Accuracy:  0.911, Validation Accuracy:  0.903, Loss:  0.058
Epoch   2 Batch  491/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.905, Loss:  0.060
Epoch   2 Batch  492/1077 - Train Accuracy:  0.944, Validation Accuracy:  0.909, Loss:  0.062
Epoch   2 Batch  493/1077 - Train Accuracy:  0.946, Validation Accuracy:  0.910, Loss:  0.033
Epoch   2 Batch  494/1077 - Train Accuracy:  0.922, Validation Accuracy:  0.910, Loss:  0.038
Epoch   2 Batch  495/1077 - Train Accuracy:  0.893, Validation Accuracy:  0.913, Loss:  0.051
Epoch   2 Batch  496/1077 - Train Accuracy:  0.923, Validation Accuracy:  0.920, Loss:  0.061
Epoch   2 Batch  497/1077 - Train Accuracy:  0.938, Validation Accuracy:  0.920, Loss:  0.058
Epoch   2 Batch  498/1077 - Train Accuracy:  0.927, Validati

Epoch   2 Batch  576/1077 - Train Accuracy:  0.960, Validation Accuracy:  0.915, Loss:  0.048
Epoch   2 Batch  577/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.922, Loss:  0.056
Epoch   2 Batch  578/1077 - Train Accuracy:  0.950, Validation Accuracy:  0.922, Loss:  0.045
Epoch   2 Batch  579/1077 - Train Accuracy:  0.935, Validation Accuracy:  0.922, Loss:  0.045
Epoch   2 Batch  580/1077 - Train Accuracy:  0.940, Validation Accuracy:  0.922, Loss:  0.042
Epoch   2 Batch  581/1077 - Train Accuracy:  0.952, Validation Accuracy:  0.922, Loss:  0.037
Epoch   2 Batch  582/1077 - Train Accuracy:  0.928, Validation Accuracy:  0.915, Loss:  0.045
Epoch   2 Batch  583/1077 - Train Accuracy:  0.938, Validation Accuracy:  0.913, Loss:  0.057
Epoch   2 Batch  584/1077 - Train Accuracy:  0.954, Validation Accuracy:  0.913, Loss:  0.050
Epoch   2 Batch  585/1077 - Train Accuracy:  0.942, Validation Accuracy:  0.912, Loss:  0.044
Epoch   2 Batch  586/1077 - Train Accuracy:  0.920, Validati

Epoch   2 Batch  664/1077 - Train Accuracy:  0.923, Validation Accuracy:  0.900, Loss:  0.052
Epoch   2 Batch  665/1077 - Train Accuracy:  0.929, Validation Accuracy:  0.899, Loss:  0.048
Epoch   2 Batch  666/1077 - Train Accuracy:  0.914, Validation Accuracy:  0.899, Loss:  0.063
Epoch   2 Batch  667/1077 - Train Accuracy:  0.924, Validation Accuracy:  0.897, Loss:  0.060
Epoch   2 Batch  668/1077 - Train Accuracy:  0.922, Validation Accuracy:  0.905, Loss:  0.042
Epoch   2 Batch  669/1077 - Train Accuracy:  0.933, Validation Accuracy:  0.915, Loss:  0.049
Epoch   2 Batch  670/1077 - Train Accuracy:  0.941, Validation Accuracy:  0.912, Loss:  0.058
Epoch   2 Batch  671/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.908, Loss:  0.059
Epoch   2 Batch  672/1077 - Train Accuracy:  0.919, Validation Accuracy:  0.909, Loss:  0.045
Epoch   2 Batch  673/1077 - Train Accuracy:  0.929, Validation Accuracy:  0.914, Loss:  0.048
Epoch   2 Batch  674/1077 - Train Accuracy:  0.936, Validati

Epoch   2 Batch  752/1077 - Train Accuracy:  0.937, Validation Accuracy:  0.942, Loss:  0.047
Epoch   2 Batch  753/1077 - Train Accuracy:  0.932, Validation Accuracy:  0.938, Loss:  0.055
Epoch   2 Batch  754/1077 - Train Accuracy:  0.914, Validation Accuracy:  0.933, Loss:  0.061
Epoch   2 Batch  755/1077 - Train Accuracy:  0.939, Validation Accuracy:  0.942, Loss:  0.062
Epoch   2 Batch  756/1077 - Train Accuracy:  0.949, Validation Accuracy:  0.942, Loss:  0.049
Epoch   2 Batch  757/1077 - Train Accuracy:  0.920, Validation Accuracy:  0.939, Loss:  0.045
Epoch   2 Batch  758/1077 - Train Accuracy:  0.946, Validation Accuracy:  0.934, Loss:  0.047
Epoch   2 Batch  759/1077 - Train Accuracy:  0.943, Validation Accuracy:  0.929, Loss:  0.045
Epoch   2 Batch  760/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.929, Loss:  0.055
Epoch   2 Batch  761/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.929, Loss:  0.049
Epoch   2 Batch  762/1077 - Train Accuracy:  0.921, Validati

Epoch   2 Batch  840/1077 - Train Accuracy:  0.959, Validation Accuracy:  0.924, Loss:  0.039
Epoch   2 Batch  841/1077 - Train Accuracy:  0.933, Validation Accuracy:  0.924, Loss:  0.050
Epoch   2 Batch  842/1077 - Train Accuracy:  0.944, Validation Accuracy:  0.926, Loss:  0.044
Epoch   2 Batch  843/1077 - Train Accuracy:  0.931, Validation Accuracy:  0.926, Loss:  0.044
Epoch   2 Batch  844/1077 - Train Accuracy:  0.923, Validation Accuracy:  0.918, Loss:  0.039
Epoch   2 Batch  845/1077 - Train Accuracy:  0.939, Validation Accuracy:  0.926, Loss:  0.039
Epoch   2 Batch  846/1077 - Train Accuracy:  0.937, Validation Accuracy:  0.930, Loss:  0.063
Epoch   2 Batch  847/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.930, Loss:  0.057
Epoch   2 Batch  848/1077 - Train Accuracy:  0.939, Validation Accuracy:  0.925, Loss:  0.045
Epoch   2 Batch  849/1077 - Train Accuracy:  0.941, Validation Accuracy:  0.922, Loss:  0.042
Epoch   2 Batch  850/1077 - Train Accuracy:  0.910, Validati

Epoch   2 Batch  928/1077 - Train Accuracy:  0.934, Validation Accuracy:  0.909, Loss:  0.046
Epoch   2 Batch  929/1077 - Train Accuracy:  0.922, Validation Accuracy:  0.906, Loss:  0.060
Epoch   2 Batch  930/1077 - Train Accuracy:  0.940, Validation Accuracy:  0.909, Loss:  0.047
Epoch   2 Batch  931/1077 - Train Accuracy:  0.941, Validation Accuracy:  0.908, Loss:  0.038
Epoch   2 Batch  932/1077 - Train Accuracy:  0.925, Validation Accuracy:  0.916, Loss:  0.050
Epoch   2 Batch  933/1077 - Train Accuracy:  0.913, Validation Accuracy:  0.917, Loss:  0.056
Epoch   2 Batch  934/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.928, Loss:  0.047
Epoch   2 Batch  935/1077 - Train Accuracy:  0.953, Validation Accuracy:  0.928, Loss:  0.044
Epoch   2 Batch  936/1077 - Train Accuracy:  0.936, Validation Accuracy:  0.928, Loss:  0.051
Epoch   2 Batch  937/1077 - Train Accuracy:  0.924, Validation Accuracy:  0.918, Loss:  0.074
Epoch   2 Batch  938/1077 - Train Accuracy:  0.934, Validati

Epoch   2 Batch 1016/1077 - Train Accuracy:  0.915, Validation Accuracy:  0.933, Loss:  0.054
Epoch   2 Batch 1017/1077 - Train Accuracy:  0.932, Validation Accuracy:  0.931, Loss:  0.055
Epoch   2 Batch 1018/1077 - Train Accuracy:  0.920, Validation Accuracy:  0.941, Loss:  0.041
Epoch   2 Batch 1019/1077 - Train Accuracy:  0.912, Validation Accuracy:  0.948, Loss:  0.074
Epoch   2 Batch 1020/1077 - Train Accuracy:  0.956, Validation Accuracy:  0.941, Loss:  0.042
Epoch   2 Batch 1021/1077 - Train Accuracy:  0.920, Validation Accuracy:  0.936, Loss:  0.057
Epoch   2 Batch 1022/1077 - Train Accuracy:  0.943, Validation Accuracy:  0.930, Loss:  0.050
Epoch   2 Batch 1023/1077 - Train Accuracy:  0.921, Validation Accuracy:  0.932, Loss:  0.052
Epoch   2 Batch 1024/1077 - Train Accuracy:  0.912, Validation Accuracy:  0.936, Loss:  0.066
Epoch   2 Batch 1025/1077 - Train Accuracy:  0.913, Validation Accuracy:  0.940, Loss:  0.055
Epoch   2 Batch 1026/1077 - Train Accuracy:  0.955, Validati

Epoch   3 Batch   28/1077 - Train Accuracy:  0.951, Validation Accuracy:  0.926, Loss:  0.053
Epoch   3 Batch   29/1077 - Train Accuracy:  0.924, Validation Accuracy:  0.925, Loss:  0.056
Epoch   3 Batch   30/1077 - Train Accuracy:  0.940, Validation Accuracy:  0.920, Loss:  0.036
Epoch   3 Batch   31/1077 - Train Accuracy:  0.931, Validation Accuracy:  0.925, Loss:  0.041
Epoch   3 Batch   32/1077 - Train Accuracy:  0.926, Validation Accuracy:  0.924, Loss:  0.045
Epoch   3 Batch   33/1077 - Train Accuracy:  0.936, Validation Accuracy:  0.918, Loss:  0.055
Epoch   3 Batch   34/1077 - Train Accuracy:  0.937, Validation Accuracy:  0.929, Loss:  0.050
Epoch   3 Batch   35/1077 - Train Accuracy:  0.966, Validation Accuracy:  0.929, Loss:  0.051
Epoch   3 Batch   36/1077 - Train Accuracy:  0.924, Validation Accuracy:  0.934, Loss:  0.043
Epoch   3 Batch   37/1077 - Train Accuracy:  0.946, Validation Accuracy:  0.934, Loss:  0.043
Epoch   3 Batch   38/1077 - Train Accuracy:  0.920, Validati

Epoch   3 Batch  116/1077 - Train Accuracy:  0.893, Validation Accuracy:  0.920, Loss:  0.066
Epoch   3 Batch  117/1077 - Train Accuracy:  0.913, Validation Accuracy:  0.923, Loss:  0.049
Epoch   3 Batch  118/1077 - Train Accuracy:  0.908, Validation Accuracy:  0.925, Loss:  0.044
Epoch   3 Batch  119/1077 - Train Accuracy:  0.914, Validation Accuracy:  0.936, Loss:  0.046
Epoch   3 Batch  120/1077 - Train Accuracy:  0.917, Validation Accuracy:  0.940, Loss:  0.058
Epoch   3 Batch  121/1077 - Train Accuracy:  0.945, Validation Accuracy:  0.936, Loss:  0.057
Epoch   3 Batch  122/1077 - Train Accuracy:  0.933, Validation Accuracy:  0.936, Loss:  0.043
Epoch   3 Batch  123/1077 - Train Accuracy:  0.925, Validation Accuracy:  0.941, Loss:  0.045
Epoch   3 Batch  124/1077 - Train Accuracy:  0.911, Validation Accuracy:  0.936, Loss:  0.073
Epoch   3 Batch  125/1077 - Train Accuracy:  0.932, Validation Accuracy:  0.934, Loss:  0.056
Epoch   3 Batch  126/1077 - Train Accuracy:  0.933, Validati

Epoch   3 Batch  204/1077 - Train Accuracy:  0.933, Validation Accuracy:  0.932, Loss:  0.074
Epoch   3 Batch  205/1077 - Train Accuracy:  0.886, Validation Accuracy:  0.932, Loss:  0.077
Epoch   3 Batch  206/1077 - Train Accuracy:  0.948, Validation Accuracy:  0.933, Loss:  0.040
Epoch   3 Batch  207/1077 - Train Accuracy:  0.929, Validation Accuracy:  0.928, Loss:  0.045
Epoch   3 Batch  208/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.928, Loss:  0.049
Epoch   3 Batch  209/1077 - Train Accuracy:  0.948, Validation Accuracy:  0.927, Loss:  0.030
Epoch   3 Batch  210/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.923, Loss:  0.060
Epoch   3 Batch  211/1077 - Train Accuracy:  0.940, Validation Accuracy:  0.922, Loss:  0.046
Epoch   3 Batch  212/1077 - Train Accuracy:  0.933, Validation Accuracy:  0.926, Loss:  0.043
Epoch   3 Batch  213/1077 - Train Accuracy:  0.944, Validation Accuracy:  0.922, Loss:  0.043
Epoch   3 Batch  214/1077 - Train Accuracy:  0.934, Validati

Epoch   3 Batch  292/1077 - Train Accuracy:  0.904, Validation Accuracy:  0.929, Loss:  0.052
Epoch   3 Batch  293/1077 - Train Accuracy:  0.934, Validation Accuracy:  0.932, Loss:  0.072
Epoch   3 Batch  294/1077 - Train Accuracy:  0.937, Validation Accuracy:  0.933, Loss:  0.042
Epoch   3 Batch  295/1077 - Train Accuracy:  0.934, Validation Accuracy:  0.931, Loss:  0.055
Epoch   3 Batch  296/1077 - Train Accuracy:  0.936, Validation Accuracy:  0.930, Loss:  0.057
Epoch   3 Batch  297/1077 - Train Accuracy:  0.921, Validation Accuracy:  0.928, Loss:  0.068
Epoch   3 Batch  298/1077 - Train Accuracy:  0.925, Validation Accuracy:  0.928, Loss:  0.065
Epoch   3 Batch  299/1077 - Train Accuracy:  0.942, Validation Accuracy:  0.924, Loss:  0.052
Epoch   3 Batch  300/1077 - Train Accuracy:  0.942, Validation Accuracy:  0.914, Loss:  0.042
Epoch   3 Batch  301/1077 - Train Accuracy:  0.928, Validation Accuracy:  0.914, Loss:  0.043
Epoch   3 Batch  302/1077 - Train Accuracy:  0.951, Validati

Epoch   3 Batch  380/1077 - Train Accuracy:  0.943, Validation Accuracy:  0.933, Loss:  0.045
Epoch   3 Batch  381/1077 - Train Accuracy:  0.919, Validation Accuracy:  0.947, Loss:  0.055
Epoch   3 Batch  382/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.947, Loss:  0.073
Epoch   3 Batch  383/1077 - Train Accuracy:  0.945, Validation Accuracy:  0.947, Loss:  0.046
Epoch   3 Batch  384/1077 - Train Accuracy:  0.942, Validation Accuracy:  0.947, Loss:  0.031
Epoch   3 Batch  385/1077 - Train Accuracy:  0.943, Validation Accuracy:  0.956, Loss:  0.046
Epoch   3 Batch  386/1077 - Train Accuracy:  0.936, Validation Accuracy:  0.954, Loss:  0.042
Epoch   3 Batch  387/1077 - Train Accuracy:  0.949, Validation Accuracy:  0.953, Loss:  0.038
Epoch   3 Batch  388/1077 - Train Accuracy:  0.928, Validation Accuracy:  0.955, Loss:  0.050
Epoch   3 Batch  389/1077 - Train Accuracy:  0.934, Validation Accuracy:  0.945, Loss:  0.047
Epoch   3 Batch  390/1077 - Train Accuracy:  0.897, Validati

Epoch   3 Batch  468/1077 - Train Accuracy:  0.936, Validation Accuracy:  0.923, Loss:  0.057
Epoch   3 Batch  469/1077 - Train Accuracy:  0.938, Validation Accuracy:  0.920, Loss:  0.061
Epoch   3 Batch  470/1077 - Train Accuracy:  0.953, Validation Accuracy:  0.917, Loss:  0.053
Epoch   3 Batch  471/1077 - Train Accuracy:  0.951, Validation Accuracy:  0.913, Loss:  0.046
Epoch   3 Batch  472/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.909, Loss:  0.049
Epoch   3 Batch  473/1077 - Train Accuracy:  0.923, Validation Accuracy:  0.909, Loss:  0.046
Epoch   3 Batch  474/1077 - Train Accuracy:  0.934, Validation Accuracy:  0.916, Loss:  0.047
Epoch   3 Batch  475/1077 - Train Accuracy:  0.921, Validation Accuracy:  0.912, Loss:  0.051
Epoch   3 Batch  476/1077 - Train Accuracy:  0.914, Validation Accuracy:  0.917, Loss:  0.039
Epoch   3 Batch  477/1077 - Train Accuracy:  0.941, Validation Accuracy:  0.925, Loss:  0.052
Epoch   3 Batch  478/1077 - Train Accuracy:  0.940, Validati

Epoch   3 Batch  556/1077 - Train Accuracy:  0.928, Validation Accuracy:  0.907, Loss:  0.052
Epoch   3 Batch  557/1077 - Train Accuracy:  0.921, Validation Accuracy:  0.906, Loss:  0.056
Epoch   3 Batch  558/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.901, Loss:  0.038
Epoch   3 Batch  559/1077 - Train Accuracy:  0.938, Validation Accuracy:  0.908, Loss:  0.046
Epoch   3 Batch  560/1077 - Train Accuracy:  0.942, Validation Accuracy:  0.904, Loss:  0.049
Epoch   3 Batch  561/1077 - Train Accuracy:  0.937, Validation Accuracy:  0.912, Loss:  0.042
Epoch   3 Batch  562/1077 - Train Accuracy:  0.931, Validation Accuracy:  0.916, Loss:  0.046
Epoch   3 Batch  563/1077 - Train Accuracy:  0.909, Validation Accuracy:  0.904, Loss:  0.057
Epoch   3 Batch  564/1077 - Train Accuracy:  0.948, Validation Accuracy:  0.901, Loss:  0.057
Epoch   3 Batch  565/1077 - Train Accuracy:  0.922, Validation Accuracy:  0.907, Loss:  0.068
Epoch   3 Batch  566/1077 - Train Accuracy:  0.930, Validati

Epoch   3 Batch  644/1077 - Train Accuracy:  0.920, Validation Accuracy:  0.925, Loss:  0.052
Epoch   3 Batch  645/1077 - Train Accuracy:  0.937, Validation Accuracy:  0.930, Loss:  0.061
Epoch   3 Batch  646/1077 - Train Accuracy:  0.938, Validation Accuracy:  0.931, Loss:  0.048
Epoch   3 Batch  647/1077 - Train Accuracy:  0.928, Validation Accuracy:  0.931, Loss:  0.057
Epoch   3 Batch  648/1077 - Train Accuracy:  0.962, Validation Accuracy:  0.931, Loss:  0.031
Epoch   3 Batch  649/1077 - Train Accuracy:  0.953, Validation Accuracy:  0.931, Loss:  0.048
Epoch   3 Batch  650/1077 - Train Accuracy:  0.933, Validation Accuracy:  0.926, Loss:  0.040
Epoch   3 Batch  651/1077 - Train Accuracy:  0.948, Validation Accuracy:  0.931, Loss:  0.044
Epoch   3 Batch  652/1077 - Train Accuracy:  0.945, Validation Accuracy:  0.936, Loss:  0.065
Epoch   3 Batch  653/1077 - Train Accuracy:  0.940, Validation Accuracy:  0.934, Loss:  0.054
Epoch   3 Batch  654/1077 - Train Accuracy:  0.945, Validati

Epoch   3 Batch  732/1077 - Train Accuracy:  0.899, Validation Accuracy:  0.918, Loss:  0.059
Epoch   3 Batch  733/1077 - Train Accuracy:  0.940, Validation Accuracy:  0.904, Loss:  0.046
Epoch   3 Batch  734/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.907, Loss:  0.054
Epoch   3 Batch  735/1077 - Train Accuracy:  0.922, Validation Accuracy:  0.911, Loss:  0.047
Epoch   3 Batch  736/1077 - Train Accuracy:  0.941, Validation Accuracy:  0.916, Loss:  0.041
Epoch   3 Batch  737/1077 - Train Accuracy:  0.933, Validation Accuracy:  0.922, Loss:  0.063
Epoch   3 Batch  738/1077 - Train Accuracy:  0.941, Validation Accuracy:  0.926, Loss:  0.039
Epoch   3 Batch  739/1077 - Train Accuracy:  0.943, Validation Accuracy:  0.928, Loss:  0.053
Epoch   3 Batch  740/1077 - Train Accuracy:  0.933, Validation Accuracy:  0.929, Loss:  0.043
Epoch   3 Batch  741/1077 - Train Accuracy:  0.938, Validation Accuracy:  0.931, Loss:  0.058
Epoch   3 Batch  742/1077 - Train Accuracy:  0.936, Validati

Epoch   3 Batch  820/1077 - Train Accuracy:  0.958, Validation Accuracy:  0.927, Loss:  0.048
Epoch   3 Batch  821/1077 - Train Accuracy:  0.937, Validation Accuracy:  0.923, Loss:  0.054
Epoch   3 Batch  822/1077 - Train Accuracy:  0.961, Validation Accuracy:  0.918, Loss:  0.040
Epoch   3 Batch  823/1077 - Train Accuracy:  0.922, Validation Accuracy:  0.922, Loss:  0.058
Epoch   3 Batch  824/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.922, Loss:  0.064
Epoch   3 Batch  825/1077 - Train Accuracy:  0.943, Validation Accuracy:  0.919, Loss:  0.043
Epoch   3 Batch  826/1077 - Train Accuracy:  0.913, Validation Accuracy:  0.923, Loss:  0.048
Epoch   3 Batch  827/1077 - Train Accuracy:  0.941, Validation Accuracy:  0.926, Loss:  0.047
Epoch   3 Batch  828/1077 - Train Accuracy:  0.937, Validation Accuracy:  0.922, Loss:  0.051
Epoch   3 Batch  829/1077 - Train Accuracy:  0.914, Validation Accuracy:  0.918, Loss:  0.069
Epoch   3 Batch  830/1077 - Train Accuracy:  0.898, Validati

Epoch   3 Batch  908/1077 - Train Accuracy:  0.928, Validation Accuracy:  0.936, Loss:  0.052
Epoch   3 Batch  909/1077 - Train Accuracy:  0.938, Validation Accuracy:  0.936, Loss:  0.052
Epoch   3 Batch  910/1077 - Train Accuracy:  0.934, Validation Accuracy:  0.935, Loss:  0.049
Epoch   3 Batch  911/1077 - Train Accuracy:  0.929, Validation Accuracy:  0.928, Loss:  0.045
Epoch   3 Batch  912/1077 - Train Accuracy:  0.938, Validation Accuracy:  0.930, Loss:  0.047
Epoch   3 Batch  913/1077 - Train Accuracy:  0.925, Validation Accuracy:  0.926, Loss:  0.078
Epoch   3 Batch  914/1077 - Train Accuracy:  0.945, Validation Accuracy:  0.927, Loss:  0.067
Epoch   3 Batch  915/1077 - Train Accuracy:  0.924, Validation Accuracy:  0.916, Loss:  0.045
Epoch   3 Batch  916/1077 - Train Accuracy:  0.920, Validation Accuracy:  0.912, Loss:  0.050
Epoch   3 Batch  917/1077 - Train Accuracy:  0.939, Validation Accuracy:  0.915, Loss:  0.047
Epoch   3 Batch  918/1077 - Train Accuracy:  0.938, Validati

Epoch   3 Batch  996/1077 - Train Accuracy:  0.926, Validation Accuracy:  0.923, Loss:  0.052
Epoch   3 Batch  997/1077 - Train Accuracy:  0.953, Validation Accuracy:  0.922, Loss:  0.045
Epoch   3 Batch  998/1077 - Train Accuracy:  0.914, Validation Accuracy:  0.926, Loss:  0.050
Epoch   3 Batch  999/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.929, Loss:  0.054
Epoch   3 Batch 1000/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.929, Loss:  0.041
Epoch   3 Batch 1001/1077 - Train Accuracy:  0.923, Validation Accuracy:  0.925, Loss:  0.035
Epoch   3 Batch 1002/1077 - Train Accuracy:  0.959, Validation Accuracy:  0.928, Loss:  0.036
Epoch   3 Batch 1003/1077 - Train Accuracy:  0.945, Validation Accuracy:  0.933, Loss:  0.050
Epoch   3 Batch 1004/1077 - Train Accuracy:  0.964, Validation Accuracy:  0.933, Loss:  0.052
Epoch   3 Batch 1005/1077 - Train Accuracy:  0.942, Validation Accuracy:  0.931, Loss:  0.043
Epoch   3 Batch 1006/1077 - Train Accuracy:  0.940, Validati

Epoch   4 Batch    8/1077 - Train Accuracy:  0.948, Validation Accuracy:  0.914, Loss:  0.041
Epoch   4 Batch    9/1077 - Train Accuracy:  0.925, Validation Accuracy:  0.912, Loss:  0.045
Epoch   4 Batch   10/1077 - Train Accuracy:  0.947, Validation Accuracy:  0.912, Loss:  0.048
Epoch   4 Batch   11/1077 - Train Accuracy:  0.911, Validation Accuracy:  0.919, Loss:  0.058
Epoch   4 Batch   12/1077 - Train Accuracy:  0.950, Validation Accuracy:  0.922, Loss:  0.046
Epoch   4 Batch   13/1077 - Train Accuracy:  0.949, Validation Accuracy:  0.931, Loss:  0.058
Epoch   4 Batch   14/1077 - Train Accuracy:  0.938, Validation Accuracy:  0.920, Loss:  0.039
Epoch   4 Batch   15/1077 - Train Accuracy:  0.952, Validation Accuracy:  0.917, Loss:  0.038
Epoch   4 Batch   16/1077 - Train Accuracy:  0.944, Validation Accuracy:  0.924, Loss:  0.060
Epoch   4 Batch   17/1077 - Train Accuracy:  0.940, Validation Accuracy:  0.924, Loss:  0.044
Epoch   4 Batch   18/1077 - Train Accuracy:  0.927, Validati

Epoch   4 Batch   96/1077 - Train Accuracy:  0.932, Validation Accuracy:  0.918, Loss:  0.045
Epoch   4 Batch   97/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.918, Loss:  0.050
Epoch   4 Batch   98/1077 - Train Accuracy:  0.920, Validation Accuracy:  0.928, Loss:  0.052
Epoch   4 Batch   99/1077 - Train Accuracy:  0.934, Validation Accuracy:  0.928, Loss:  0.043
Epoch   4 Batch  100/1077 - Train Accuracy:  0.932, Validation Accuracy:  0.929, Loss:  0.048
Epoch   4 Batch  101/1077 - Train Accuracy:  0.943, Validation Accuracy:  0.929, Loss:  0.048
Epoch   4 Batch  102/1077 - Train Accuracy:  0.946, Validation Accuracy:  0.931, Loss:  0.042
Epoch   4 Batch  103/1077 - Train Accuracy:  0.926, Validation Accuracy:  0.940, Loss:  0.049
Epoch   4 Batch  104/1077 - Train Accuracy:  0.940, Validation Accuracy:  0.925, Loss:  0.054
Epoch   4 Batch  105/1077 - Train Accuracy:  0.933, Validation Accuracy:  0.929, Loss:  0.047
Epoch   4 Batch  106/1077 - Train Accuracy:  0.958, Validati

Epoch   4 Batch  184/1077 - Train Accuracy:  0.938, Validation Accuracy:  0.904, Loss:  0.050
Epoch   4 Batch  185/1077 - Train Accuracy:  0.914, Validation Accuracy:  0.911, Loss:  0.060
Epoch   4 Batch  186/1077 - Train Accuracy:  0.942, Validation Accuracy:  0.909, Loss:  0.053
Epoch   4 Batch  187/1077 - Train Accuracy:  0.916, Validation Accuracy:  0.908, Loss:  0.038
Epoch   4 Batch  188/1077 - Train Accuracy:  0.943, Validation Accuracy:  0.908, Loss:  0.049
Epoch   4 Batch  189/1077 - Train Accuracy:  0.941, Validation Accuracy:  0.909, Loss:  0.040
Epoch   4 Batch  190/1077 - Train Accuracy:  0.945, Validation Accuracy:  0.916, Loss:  0.048
Epoch   4 Batch  191/1077 - Train Accuracy:  0.907, Validation Accuracy:  0.924, Loss:  0.039
Epoch   4 Batch  192/1077 - Train Accuracy:  0.954, Validation Accuracy:  0.925, Loss:  0.049
Epoch   4 Batch  193/1077 - Train Accuracy:  0.954, Validation Accuracy:  0.931, Loss:  0.043
Epoch   4 Batch  194/1077 - Train Accuracy:  0.946, Validati

Epoch   4 Batch  272/1077 - Train Accuracy:  0.931, Validation Accuracy:  0.932, Loss:  0.080
Epoch   4 Batch  273/1077 - Train Accuracy:  0.924, Validation Accuracy:  0.941, Loss:  0.046
Epoch   4 Batch  274/1077 - Train Accuracy:  0.911, Validation Accuracy:  0.936, Loss:  0.057
Epoch   4 Batch  275/1077 - Train Accuracy:  0.923, Validation Accuracy:  0.936, Loss:  0.050
Epoch   4 Batch  276/1077 - Train Accuracy:  0.934, Validation Accuracy:  0.924, Loss:  0.080
Epoch   4 Batch  277/1077 - Train Accuracy:  0.936, Validation Accuracy:  0.926, Loss:  0.038
Epoch   4 Batch  278/1077 - Train Accuracy:  0.911, Validation Accuracy:  0.926, Loss:  0.067
Epoch   4 Batch  279/1077 - Train Accuracy:  0.943, Validation Accuracy:  0.924, Loss:  0.063
Epoch   4 Batch  280/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.925, Loss:  0.058
Epoch   4 Batch  281/1077 - Train Accuracy:  0.949, Validation Accuracy:  0.926, Loss:  0.059
Epoch   4 Batch  282/1077 - Train Accuracy:  0.928, Validati

Epoch   4 Batch  360/1077 - Train Accuracy:  0.950, Validation Accuracy:  0.916, Loss:  0.039
Epoch   4 Batch  361/1077 - Train Accuracy:  0.948, Validation Accuracy:  0.919, Loss:  0.055
Epoch   4 Batch  362/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.924, Loss:  0.051
Epoch   4 Batch  363/1077 - Train Accuracy:  0.910, Validation Accuracy:  0.928, Loss:  0.059
Epoch   4 Batch  364/1077 - Train Accuracy:  0.915, Validation Accuracy:  0.923, Loss:  0.057
Epoch   4 Batch  365/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.929, Loss:  0.044
Epoch   4 Batch  366/1077 - Train Accuracy:  0.949, Validation Accuracy:  0.925, Loss:  0.044
Epoch   4 Batch  367/1077 - Train Accuracy:  0.949, Validation Accuracy:  0.915, Loss:  0.040
Epoch   4 Batch  368/1077 - Train Accuracy:  0.946, Validation Accuracy:  0.913, Loss:  0.051
Epoch   4 Batch  369/1077 - Train Accuracy:  0.942, Validation Accuracy:  0.909, Loss:  0.048
Epoch   4 Batch  370/1077 - Train Accuracy:  0.943, Validati

Epoch   4 Batch  448/1077 - Train Accuracy:  0.921, Validation Accuracy:  0.909, Loss:  0.067
Epoch   4 Batch  449/1077 - Train Accuracy:  0.918, Validation Accuracy:  0.905, Loss:  0.066
Epoch   4 Batch  450/1077 - Train Accuracy:  0.952, Validation Accuracy:  0.916, Loss:  0.045
Epoch   4 Batch  451/1077 - Train Accuracy:  0.939, Validation Accuracy:  0.910, Loss:  0.053
Epoch   4 Batch  452/1077 - Train Accuracy:  0.910, Validation Accuracy:  0.911, Loss:  0.054
Epoch   4 Batch  453/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.907, Loss:  0.049
Epoch   4 Batch  454/1077 - Train Accuracy:  0.932, Validation Accuracy:  0.911, Loss:  0.057
Epoch   4 Batch  455/1077 - Train Accuracy:  0.929, Validation Accuracy:  0.910, Loss:  0.064
Epoch   4 Batch  456/1077 - Train Accuracy:  0.939, Validation Accuracy:  0.916, Loss:  0.061
Epoch   4 Batch  457/1077 - Train Accuracy:  0.928, Validation Accuracy:  0.912, Loss:  0.050
Epoch   4 Batch  458/1077 - Train Accuracy:  0.907, Validati

Epoch   4 Batch  536/1077 - Train Accuracy:  0.887, Validation Accuracy:  0.911, Loss:  0.057
Epoch   4 Batch  537/1077 - Train Accuracy:  0.926, Validation Accuracy:  0.922, Loss:  0.041
Epoch   4 Batch  538/1077 - Train Accuracy:  0.948, Validation Accuracy:  0.921, Loss:  0.041
Epoch   4 Batch  539/1077 - Train Accuracy:  0.893, Validation Accuracy:  0.922, Loss:  0.070
Epoch   4 Batch  540/1077 - Train Accuracy:  0.928, Validation Accuracy:  0.918, Loss:  0.052
Epoch   4 Batch  541/1077 - Train Accuracy:  0.916, Validation Accuracy:  0.917, Loss:  0.061
Epoch   4 Batch  542/1077 - Train Accuracy:  0.933, Validation Accuracy:  0.913, Loss:  0.056
Epoch   4 Batch  543/1077 - Train Accuracy:  0.936, Validation Accuracy:  0.914, Loss:  0.056
Epoch   4 Batch  544/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.914, Loss:  0.051
Epoch   4 Batch  545/1077 - Train Accuracy:  0.922, Validation Accuracy:  0.908, Loss:  0.056
Epoch   4 Batch  546/1077 - Train Accuracy:  0.922, Validati

Epoch   4 Batch  624/1077 - Train Accuracy:  0.939, Validation Accuracy:  0.935, Loss:  0.053
Epoch   4 Batch  625/1077 - Train Accuracy:  0.936, Validation Accuracy:  0.934, Loss:  0.047
Epoch   4 Batch  626/1077 - Train Accuracy:  0.925, Validation Accuracy:  0.934, Loss:  0.057
Epoch   4 Batch  627/1077 - Train Accuracy:  0.952, Validation Accuracy:  0.934, Loss:  0.049
Epoch   4 Batch  628/1077 - Train Accuracy:  0.933, Validation Accuracy:  0.931, Loss:  0.054
Epoch   4 Batch  629/1077 - Train Accuracy:  0.916, Validation Accuracy:  0.931, Loss:  0.064
Epoch   4 Batch  630/1077 - Train Accuracy:  0.949, Validation Accuracy:  0.931, Loss:  0.052
Epoch   4 Batch  631/1077 - Train Accuracy:  0.942, Validation Accuracy:  0.942, Loss:  0.050
Epoch   4 Batch  632/1077 - Train Accuracy:  0.950, Validation Accuracy:  0.946, Loss:  0.043
Epoch   4 Batch  633/1077 - Train Accuracy:  0.925, Validation Accuracy:  0.941, Loss:  0.048
Epoch   4 Batch  634/1077 - Train Accuracy:  0.943, Validati

Epoch   4 Batch  712/1077 - Train Accuracy:  0.946, Validation Accuracy:  0.925, Loss:  0.041
Epoch   4 Batch  713/1077 - Train Accuracy:  0.944, Validation Accuracy:  0.924, Loss:  0.049
Epoch   4 Batch  714/1077 - Train Accuracy:  0.936, Validation Accuracy:  0.928, Loss:  0.048
Epoch   4 Batch  715/1077 - Train Accuracy:  0.920, Validation Accuracy:  0.928, Loss:  0.049
Epoch   4 Batch  716/1077 - Train Accuracy:  0.960, Validation Accuracy:  0.932, Loss:  0.041
Epoch   4 Batch  717/1077 - Train Accuracy:  0.945, Validation Accuracy:  0.927, Loss:  0.044
Epoch   4 Batch  718/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.934, Loss:  0.047
Epoch   4 Batch  719/1077 - Train Accuracy:  0.905, Validation Accuracy:  0.939, Loss:  0.063
Epoch   4 Batch  720/1077 - Train Accuracy:  0.938, Validation Accuracy:  0.939, Loss:  0.053
Epoch   4 Batch  721/1077 - Train Accuracy:  0.930, Validation Accuracy:  0.939, Loss:  0.065
Epoch   4 Batch  722/1077 - Train Accuracy:  0.909, Validati

Epoch   4 Batch  800/1077 - Train Accuracy:  0.937, Validation Accuracy:  0.933, Loss:  0.049
Epoch   4 Batch  801/1077 - Train Accuracy:  0.936, Validation Accuracy:  0.935, Loss:  0.052
Epoch   4 Batch  802/1077 - Train Accuracy:  0.947, Validation Accuracy:  0.931, Loss:  0.052
Epoch   4 Batch  803/1077 - Train Accuracy:  0.948, Validation Accuracy:  0.942, Loss:  0.045
Epoch   4 Batch  804/1077 - Train Accuracy:  0.947, Validation Accuracy:  0.929, Loss:  0.038
Epoch   4 Batch  805/1077 - Train Accuracy:  0.946, Validation Accuracy:  0.936, Loss:  0.047
Epoch   4 Batch  806/1077 - Train Accuracy:  0.938, Validation Accuracy:  0.934, Loss:  0.045
Epoch   4 Batch  807/1077 - Train Accuracy:  0.924, Validation Accuracy:  0.934, Loss:  0.043
Epoch   4 Batch  808/1077 - Train Accuracy:  0.907, Validation Accuracy:  0.932, Loss:  0.062
Epoch   4 Batch  809/1077 - Train Accuracy:  0.918, Validation Accuracy:  0.928, Loss:  0.073
Epoch   4 Batch  810/1077 - Train Accuracy:  0.953, Validati

Epoch   4 Batch  888/1077 - Train Accuracy:  0.949, Validation Accuracy:  0.910, Loss:  0.040
Epoch   4 Batch  889/1077 - Train Accuracy:  0.940, Validation Accuracy:  0.913, Loss:  0.048
Epoch   4 Batch  890/1077 - Train Accuracy:  0.950, Validation Accuracy:  0.916, Loss:  0.050
Epoch   4 Batch  891/1077 - Train Accuracy:  0.919, Validation Accuracy:  0.913, Loss:  0.042
Epoch   4 Batch  892/1077 - Train Accuracy:  0.950, Validation Accuracy:  0.909, Loss:  0.042
Epoch   4 Batch  893/1077 - Train Accuracy:  0.952, Validation Accuracy:  0.910, Loss:  0.048
Epoch   4 Batch  894/1077 - Train Accuracy:  0.945, Validation Accuracy:  0.908, Loss:  0.039
Epoch   4 Batch  895/1077 - Train Accuracy:  0.939, Validation Accuracy:  0.921, Loss:  0.044
Epoch   4 Batch  896/1077 - Train Accuracy:  0.935, Validation Accuracy:  0.926, Loss:  0.052
Epoch   4 Batch  897/1077 - Train Accuracy:  0.924, Validation Accuracy:  0.929, Loss:  0.043
Epoch   4 Batch  898/1077 - Train Accuracy:  0.920, Validati

Epoch   4 Batch  976/1077 - Train Accuracy:  0.951, Validation Accuracy:  0.931, Loss:  0.038
Epoch   4 Batch  977/1077 - Train Accuracy:  0.957, Validation Accuracy:  0.934, Loss:  0.032
Epoch   4 Batch  978/1077 - Train Accuracy:  0.937, Validation Accuracy:  0.938, Loss:  0.048
Epoch   4 Batch  979/1077 - Train Accuracy:  0.919, Validation Accuracy:  0.934, Loss:  0.042
Epoch   4 Batch  980/1077 - Train Accuracy:  0.941, Validation Accuracy:  0.935, Loss:  0.046
Epoch   4 Batch  981/1077 - Train Accuracy:  0.912, Validation Accuracy:  0.935, Loss:  0.050
Epoch   4 Batch  982/1077 - Train Accuracy:  0.958, Validation Accuracy:  0.930, Loss:  0.041
Epoch   4 Batch  983/1077 - Train Accuracy:  0.933, Validation Accuracy:  0.939, Loss:  0.047
Epoch   4 Batch  984/1077 - Train Accuracy:  0.903, Validation Accuracy:  0.935, Loss:  0.050
Epoch   4 Batch  985/1077 - Train Accuracy:  0.944, Validation Accuracy:  0.935, Loss:  0.044
Epoch   4 Batch  986/1077 - Train Accuracy:  0.938, Validati

Epoch   4 Batch 1064/1077 - Train Accuracy:  0.951, Validation Accuracy:  0.940, Loss:  0.050
Epoch   4 Batch 1065/1077 - Train Accuracy:  0.938, Validation Accuracy:  0.935, Loss:  0.050
Epoch   4 Batch 1066/1077 - Train Accuracy:  0.940, Validation Accuracy:  0.929, Loss:  0.032
Epoch   4 Batch 1067/1077 - Train Accuracy:  0.927, Validation Accuracy:  0.927, Loss:  0.059
Epoch   4 Batch 1068/1077 - Train Accuracy:  0.937, Validation Accuracy:  0.927, Loss:  0.041
Epoch   4 Batch 1069/1077 - Train Accuracy:  0.953, Validation Accuracy:  0.923, Loss:  0.038
Epoch   4 Batch 1070/1077 - Train Accuracy:  0.936, Validation Accuracy:  0.924, Loss:  0.041
Epoch   4 Batch 1071/1077 - Train Accuracy:  0.917, Validation Accuracy:  0.924, Loss:  0.042
Epoch   4 Batch 1072/1077 - Train Accuracy:  0.958, Validation Accuracy:  0.924, Loss:  0.045
Epoch   4 Batch 1073/1077 - Train Accuracy:  0.945, Validation Accuracy:  0.925, Loss:  0.053
Epoch   4 Batch 1074/1077 - Train Accuracy:  0.935, Validati

### 保存参数

保存 `batch_size` 和 `save_path` 参数以进行推论（for inference）。

In [17]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# 检查点

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## 句子到序列

要向模型提供要翻译的句子，你首先需要预处理该句子。实现函数 `sentence_to_seq()` 以预处理新的句子。

- 将句子转换为小写形式
- 使用 `vocab_to_int` 将单词转换为 id
 - 如果单词不在词汇表中，将其转换为`<UNK>` 单词 id

In [19]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    word_ids = [vocab_to_int.get(word, vocab_to_int["<UNK>"]) for word in sentence.lower().split()]
    
    return word_ids


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## 翻译

将 `translate_sentence` 从英语翻译成法语。

In [20]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('logits:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence], keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in np.argmax(translate_logits, 1)]))
print('  French Words: {}'.format([target_int_to_vocab[i] for i in np.argmax(translate_logits, 1)]))

Input
  Word Ids:      [226, 218, 43, 30, 138, 202, 171]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [286, 17, 192, 281, 138, 94, 89, 60, 1]
  French Words: ['il', 'a', 'vu', 'un', 'petit', 'camion', 'jaune', '.', '<EOS>']


## 不完美的翻译

你可能注意到了，某些句子的翻译质量比其他的要好。因为你使用的数据集只有 227 个英语单词，但实际生活中有数千个单词，只有使用这些单词的句子结果才会比较理想。对于此项目，不需要达到完美的翻译。但是，如果你想创建更好的翻译模型，则需要更好的数据。

你可以使用 [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar) 语料库训练模型。该数据集拥有更多的词汇，讨论的话题也更丰富。但是，训练时间要好多天的时间，所以确保你有 GPU 并且对于我们提供的数据集，你的神经网络性能很棒。提交此项目后，别忘了研究下 WMT10 语料库。


## 提交项目

提交项目时，确保先运行所有单元，然后再保存记事本。保存记事本文件为 “dlnd_language_translation.ipynb”，再通过菜单中的“文件” ->“下载为”将其另存为 HTML 格式。提交的项目文档中需包含“helper.py”和“problem_unittests.py”文件。
